src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -158,106 +158,106 @@ AcquireMetadataLocks(List *taskList)
  * UPDATE/DELETE/UPSERT commands and exclusive locks are unnecessary.
  */
 static void
-AcquireExecutorShardLock(Task *task, CmdType commandType)
+AcquireExecutorShardLock(Task *task, ModifyLevel modLevel)
 {
-	LOCKMODE lockMode = NoLock;
 	int64 shardId = task->anchorShardId;
 
-	if (commandType == CMD_SELECT)
+	if (shardId != INVALID_SHARD_ID)
```

Comment:
Could we do an early return if `shardId == INVALID_SHARD_ID`? Long code blocks are hard to follow and this makes it seem like we may have a task without a shard ID that does have a relevant `relationRowLockList` or for which `RequiresConsistentSnapshot` is true, which I don't think can be the case.

Reply:
Discussed this offline, it seems we may want to leverage `relationRowLockList` for CTEs.